flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/BufferBuilder.java

```diff
@@ -21,62 +21,154 @@
+	 * space in the underlying {@link MemorySegment}
+	 *
 	 * @return number of copied bytes
 	 */
 	public int append(ByteBuffer source) {
-		checkState(!built);
+		checkState(!isFinished());
 
 		int needed = source.remaining();
-		int available = limit() - position;
+		int available = getMaxCapacity() - positionMarker.getCached();
 		int toCopy = Math.min(needed, available);
 
-		memorySegment.put(position, source, toCopy);
-		position += toCopy;
+		memorySegment.put(positionMarker.getCached(), source, toCopy);
+		positionMarker.move(toCopy);
 		return toCopy;
 	}
 
-	public boolean isFull() {
-		checkState(position <= limit());
-		return position == limit();
+	/**
+	 * Mark this {@link BufferBuilder} and associated {@link BufferConsumer} as finished - no new data writes will be
+	 * allowed.
+	 *
+	 * @return number of written bytes.
+	 */
+	public int finish() {
+		checkState(!isFinished());
+		positionMarker.markFinished();
+		return getWrittenBytes();
+	}
+
+	public boolean isFinished() {
+		return positionMarker.isFinished();
 	}
 
-	public Buffer build() {
-		checkState(!built);
-		built = true;
-		return new NetworkBuffer(memorySegment, recycler, true, position);
+	public boolean isFull() {
+		checkState(positionMarker.getCached() <= getMaxCapacity());
+		return positionMarker.getCached() == getMaxCapacity();
 	}
 
 	public boolean isEmpty() {
-		return position == 0;
+		return positionMarker.getCached() == 0;
 	}
 
-	private int limit() {
+	public int getMaxCapacity() {
 		return memorySegment.size();
 	}
+
+	public int getWrittenBytes() {
+		return positionMarker.getCached();
+	}
+
+	/**
+	 * Holds a reference to the current writer position. Negative values indicate that writer ({@link BufferBuilder}
+	 * has finished. Value {@code Integer.MIN_VALUE} represents finished empty buffer.
+	 */
+	@ThreadSafe
+	interface PositionMarker {
+		int FINISHED_EMPTY = Integer.MIN_VALUE;
+
+		int get();
+
+		static boolean isFinished(int position) {
+			return position < 0;
+		}
+
+		static int getAbsolute(int position) {
+			if (position == FINISHED_EMPTY) {
+				return 0;
+			}
+			return Math.abs(position);
+		}
+	}
+
+	private static class SettablePositionMarker implements PositionMarker {
+		private volatile int position = 0;
+
+		/**
+		 * Locally cached value of volatile {@code position} to avoid unnecessary volatile accesses.
+		 */
+		private int cachedPosition = 0;
+
+		@Override
+		public int get() {
+			return position;
+		}
+
+		public boolean isFinished() {
+			return PositionMarker.isFinished(cachedPosition);
+		}
+
+		public int getCached() {
+			return PositionMarker.getAbsolute(cachedPosition);
+		}
+
+		public void markFinished() {
+			int newValue = -getCached();
+			if (newValue == 0) {
+				newValue = FINISHED_EMPTY;
+			}
+			set(newValue);
+		}
+
+		public void move(int offset) {
+			set(cachedPosition + offset);
+		}
+
+		public void set(int value) {
```

Comment:
Looks like this is never called from outside, so you could call it `publishPosition()` or something similar and only do `position = cachedPosition`?

Reply:
? Would you like me to split this method into two?